### PR TITLE
Transition to bloc

### PR DESCRIPTION
**Pull request by ThinkDigitalSoftware**
added PackageDetailsBloc, and SearchBloc

### CODE REVIEW

File: lib/main.dart [hunk 1]
```diff
@@ -1,18 +1,39 @@
...
+      BlocProvider<PackageDetailsBloc>(
+        builder: (BuildContext context) =>
+            PackageDetailsBloc(client: _htmlParsingClient),
+      ),
+    ],
+    child: PubDevClientApp(),
+  ));
 }
 
 class PubDevClientApp extends StatefulWidget {
```
Comment (felangel) [inline lib/main.dart]: can we just make this a `StatelessWidget`?

File: lib/screens/home/home_bloc.dart [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
 import 'package:tavern/screens/bloc.dart';
+import 'package:tavern/src/cache.dart';
+import 'package:tavern/src/repository.dart';
 
-class HomeBloc extends Bloc<HomeEvent, HomeState> {
+class HomeBloc extends HydratedBloc<HomeEvent, HomeState> {
   final PubHtmlParsingClient client;
-
   @override
   HomeState get initialState => InitialHomeState();
```
Comment (felangel) [inline lib/screens/home/home_bloc.dart]: since you refactored this to be a Hydrated Bloc you need to update this to:
```dart
@override
HomeState get initialState => super.initialState ?? InitialHomeState();
```
Otherwise the bloc won't used the cached state when it's available.

File: lib/screens/home/home_bloc.dart [hunk 3]
```diff
@@ -69,6 +77,12 @@ class HomeBloc extends Bloc<HomeEvent, HomeState> {
     }
     return HomePreferences(sortType: sortType, filterType: filterType);
   }
+
+  @override
+  HomeState fromJson(Map<String, dynamic> json) => HomeState.fromJson(json);
```
Comment (felangel) [inline lib/screens/home/home_bloc.dart]: you might want to wrap this in a try/catch in case anything goes wrong:
```dart
@override
HomeState fromJson(Map<String, dynamic> json) {
  try {
    return HomeState.fromJson(json);
  } catch (_) {
    return null;
  }
}
```

File: lib/screens/home/home_bloc.dart [hunk 4]
```diff
@@ -69,6 +77,12 @@ class HomeBloc extends Bloc<HomeEvent, HomeState> {
...
+
+  @override
+  Map<String, dynamic> toJson(HomeState state) => state.toJson();
```
Comment (felangel) [inline lib/screens/home/home_bloc.dart]: same here:

```dart
@override
Map<String, dynamic> toJson(HomeState state) {
  try {
    return state.toJson();
  } catch (_) {
    return null;
  }
}
```

File: lib/screens/home/home_bloc.dart [hunk 5]
```diff
@@ -19,21 +21,27 @@ class HomeBloc extends Bloc<HomeEvent, HomeState> {
...
   @override
-  Stream<HomeState> mapEventToState(HomeEvent event,) async* {
+  Stream<HomeState> mapEventToState(HomeEvent event) async* {
     if (event is GetPageOfPackagesEvent) {
-      Page page = await client.getPageOfPackages(
-        pageNumber: event.pageNumber,
-        sortBy: event.sortBy,
-        filterBy: event.filterBy,
-      );
+      Page page;
```
Comment (felangel) [inline lib/screens/home/home_bloc.dart:30]: since this is a `HydratedBloc` you can now only update if the currentState doesn't have data

```dart
if (currentState.page != null) return; // don't re-fetch if we already have a cached page.
```

File: lib/screens/home/home_bloc.dart [hunk 6]
```diff
@@ -78,7 +92,56 @@ class HomePreferences {
...
+          sortType == other.sortType &&
+          filterType == other.filterType &&
+          pageNumber == other.pageNumber;
+
+  @override
+  int get hashCode =>
+      sortType.hashCode ^ filterType.hashCode ^ pageNumber.hashCode;
+}
+
+class PageCache<PageQuery, Page> extends Cache {}
```
Comment (felangel) [inline lib/screens/home/home_bloc.dart:138]: based on the optimizations you can make to the bloc, I don't think this cache is even necessary.

File: lib/screens/package_details/package_details_bloc.dart [hunk 7]
```diff
@@ -0,0 +1,56 @@
+import 'dart:async';
+
+import 'package:bloc/bloc.dart';
+import 'package:flutter/foundation.dart';
+import 'package:pub_client/pub_client.dart';
+import 'package:tavern/screens/bloc.dart';
+import 'package:tavern/screens/package_details/package_details_event.dart';
+import 'package:tavern/src/cache.dart';
+
+class PackageDetailsBloc
```
Comment (felangel) [inline lib/screens/package_details/package_details_bloc.dart:10]: you can make this a HydratedBloc as well and you don't need the PackageRepository to have it's own cache.

File: lib/screens/package_details/package_details_bloc.dart [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+  @override
+  Stream<PackageDetailsState> mapEventToState(
+    PackageDetailsEvent event,
+  ) async* {
+    if (event is GetPackageDetailsEvent) {
+      FullPackage package = await _packageRepository.get(event.packageName);
+
+      yield PackageDetailsState(package: package);
+      event.onComplete.complete();
```
Comment (felangel) [inline lib/screens/package_details/package_details_bloc.dart:29]: why is this necessary?

File: lib/screens/package_details/package_details_event.dart [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+
+import 'package:meta/meta.dart';
+
+@immutable
+abstract class PackageDetailsEvent {}
+
+class GetPackageDetailsEvent extends PackageDetailsEvent {
+  final String packageName;
+  final int packageScore;
+  final Completer onComplete = Completer();
```
Comment (felangel) [inline lib/screens/package_details/package_details_event.dart:11]: why is this needed? You can just use BlocListener to know when a state change has occurred.

File: lib/widgets/package_list_view.dart [hunk 10]
```diff
@@ -1,25 +1,60 @@
...
-      delegate: SliverChildBuilderDelegate(
-        (context, index) {
-          return PackageTile(
-            page: page,
-            index: index,
-          );
-        },
-        childCount: page.packages.length,
-      ),
+    return PagewiseSliverList<Package>(
```
Comment (felangel) [inline lib/widgets/package_list_view.dart:27]: you can refactor this to not need a Future and just use BlocBuilder to re-render when a new state is received

File: lib/widgets/package_tile.dart [hunk 11]
```diff
@@ -50,14 +49,20 @@ class PackageTile extends StatelessWidget {
...
-        arguments: PackageDetailsArguments(
-          page.packages[index].name,
-          page.packages[index].score.toString(),
-        ),
-      ),
+      onTap: () async {
+        final getPackageDetailsEvent = GetPackageDetailsEvent(
+            packageName: packageName, packageScore: packageScore);
+        BlocProvider.of<PackageDetailsBloc>(context)
+          ..dispatch(Initialize())..dispatch(getPackageDetailsEvent);
```
Comment (felangel) [inline lib/widgets/package_tile.dart]: why do you need to dispatch `Initialized()`? Shouldn't the bloc already be initialized when it's first created in BlocProvider?

**Review by felangel — CHANGES_REQUESTED** [review]
see comments

File: lib/screens/package_details/package_details_event.dart (hunk 9 above)
Comment (ThinkDigitalSoftware) [reply lib/screens/package_details/package_details_event.dart:11]: The specific custom widget it's used for needs a future and can't get its data from the bloc state,

File: lib/widgets/package_list_view.dart (hunk 10 above)
Comment (ThinkDigitalSoftware) [reply lib/widgets/package_list_view.dart:27]: I don't understand how when the widget requires a future. I'd have to fork and restructure the custom widget to do so

File: lib/widgets/package_tile.dart (hunk 11 above)
Comment (ThinkDigitalSoftware) [reply lib/widgets/package_tile.dart]: Initialize is really a call to get the first page. Since my bloc relies on async operations for data , I need my initial state to hold the first page. I can't do that with initial state. Any suggestions?

File: lib/screens/package_details/package_details_bloc.dart (hunk 8 above)
Comment (ThinkDigitalSoftware) [reply lib/screens/package_details/package_details_bloc.dart:29]: To notify the widget that's waiting for the completed future. I'm making my own future since it's required

File: lib/widgets/package_tile.dart (hunk 11 above)
Comment (felangel) [reply lib/widgets/package_tile.dart]: that's fine but I would suggest moving the `..dispatch(Initialized())` to the BlocProvider that creates the `PackageDetailsBloc` like:
```dart
BlocProvider(
  builder: (context) => PackageDetailsBloc(...)..dispatch(Initialized()),
  child: ...
)
```

File: lib/screens/package_details/package_details_bloc.dart (hunk 7 above)
Comment (ThinkDigitalSoftware) [reply lib/screens/package_details/package_details_bloc.dart:10]: I don't currently have any state to serialize the cache is going to be session based

Comment (felangel) [reply lib/screens/package_details/package_details_bloc.dart:10]: what do you mean? The PackageDetailsBloc has to have state

File: lib/main.dart (hunk 1 above)
Comment (ThinkDigitalSoftware) [reply lib/main.dart]: Done now, good eye, thanks

File: lib/screens/home/home_bloc.dart (hunk 2 above)
Comment (ThinkDigitalSoftware) [reply lib/screens/home/home_bloc.dart]: Ok, got it. Done

File: lib/screens/home/home_bloc.dart (hunk 5 above)
Comment (ThinkDigitalSoftware) [reply lib/screens/home/home_bloc.dart:30]: I don't understand. Update what?

File: lib/screens/package_details/package_details_bloc.dart (hunk 7 above)
Comment (ThinkDigitalSoftware) [reply lib/screens/package_details/package_details_bloc.dart:10]: I meant that I don't have anything I want to persist to the device, so I don't feel that I need a hydrated bloc

File: lib/widgets/package_tile.dart (hunk 11 above)
Comment (ThinkDigitalSoftware) [reply lib/widgets/package_tile.dart]: OK, I'll take a look at that

File: lib/screens/home/home_bloc.dart (hunk 3 above)
Comment (ThinkDigitalSoftware) [reply lib/screens/home/home_bloc.dart]: Done

File: lib/screens/home/home_bloc.dart (hunk 4 above)
Comment (ThinkDigitalSoftware) [reply lib/screens/home/home_bloc.dart]: Done
